infrastructure/io/src/main/java/tech/pegasys/teku/infrastructure/io/SyncDataAccessor.java

```diff
@@ -15,18 +15,59 @@
 
 import java.io.File;
 import java.io.IOException;
+import java.nio.file.AtomicMoveNotSupportedException;
 import java.nio.file.Files;
 import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.StandardCopyOption;
 import java.nio.file.StandardOpenOption;
 import java.util.Optional;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
 import org.apache.tuweni.bytes.Bytes;
+import tech.pegasys.teku.infrastructure.exceptions.InvalidConfigurationException;
 
 /**
  * A wrapper class for reading and writing files, ensuring that the contents is immediately flushed
  * to disk.
  */
 public class SyncDataAccessor {
 
+  private static final Logger LOG = LogManager.getLogger();
+  private boolean atomicFileMoveSupport;
+
+  SyncDataAccessor(final boolean atomicFileMoveSupport) {
+    this.atomicFileMoveSupport = atomicFileMoveSupport;
+  }
+
+  public static SyncDataAccessor create(final Path path) {
+    boolean atomicFileMoveSupport = false;
+    final Path absolutePath = path.toAbsolutePath();
+    final Path tmpFile = Paths.get(absolutePath.toString(), "_temp.tmp");
+    final Path tmpDstFile = Paths.get(absolutePath.toString(), "__temp.tmp");
+
+    try {
+      nonAtomicSyncedWrite(tmpFile, Bytes.fromHexString("0x00"));
+
+      Files.move(
+          tmpFile, tmpDstFile, StandardCopyOption.ATOMIC_MOVE, StandardCopyOption.REPLACE_EXISTING);
+      atomicFileMoveSupport = true;
+    } catch (AtomicMoveNotSupportedException e) {
+      LOG.debug("File system doesn't support atomic move");
+      atomicFileMoveSupport = false;
```

Comment:
There's a bunch of duplication here with `atomicSyncWrite`.  I'd be tempted to have `atomicSyncWrite` just throw the `AtomicMoveNotSupportedException` and then this could just call `atomicSyncWrite(tmpFile, Bytes.ZERO)` to test if it worked.

Given we're doing feature detection ahead of time it's probably reasonable to just fail if the atomic move suddenly starts failing later. That would indicate something very strange is going on and we probably should stop signing stuff until it gets sorted out anyway.